crates/stackable-versioned-macros/src/codegen/container/struct/k8s.rs

```diff
@@ -173,21 +174,37 @@ impl Struct {
                     #k8s_openapi_path::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition,
                     #kube_core_path::crd::MergeError>
                 {
-                    #kube_core_path::crd::merge_crds(vec![#(#crd_fns),*], stored_apiversion.as_str())
+                    #kube_core_path::crd::merge_crds(vec![#(#crd_fns),*], stored_apiversion.as_version_str())
                 }
 
                 #convert_method
 
-                fn from_json_value(value: #serde_json_path::Value) -> ::std::result::Result<Self, #parse_object_error> {
-                    let api_version = value
+                fn from_json_value(object: #serde_json_path::Value) -> ::std::result::Result<Self, #parse_object_error> {
+                    let object_kind = object
+                        .get("kind")
+                        .ok_or_else(|| #parse_object_error::FieldMissing{ field: "kind".to_string() })?
+                        .as_str()
+                        .ok_or_else(|| #parse_object_error::FieldNotStr{ field: "kind".to_string() })?;
+
+                    // Note(@sbernauer): One could argue we don't need to check the kind, but this
+                    // is a nice sanity check. If for *some* reason a unexpected kind ends up at a
+                    // conversion, the problem might be very hard to spot without this.
```

Comment:
I think this should be rewritten to only explain why we do the check and that there is no good reason to remove it. Eg:

```suggestion
                    // Note(@sbernauer): The kind must be checked here, because it is
                    // possible for the wrong object to be deserialized.
                    // Checking here stops us assuming the kind is correct and
                    // accidentally updating upgrade/downgrade information in the
                    // status in a later step.
```

Reply:
While we check the _kind_, we should also check the _group_ matches.

That might take a bit more effort since we'd need to parse it from the `apiVersion` (could split_once on `/`). Do we assert that all versioned things have a group?